### PR TITLE
i18n for Admin Panel Settings section (WIP)

### PR DESCRIPTION
**Pull request by mazameli**
Marking as WIP because there are some clojure files like `metabot.clj` that contain strings that I'm unsure how to deal with, such as `(keys-description (format "Here's what I can %s:" verb) fn-map))`. 

Also not 100% I've done the indenting/formatting for clojure files correctly.

### CODE REVIEW

File: src/metabase/metabot.clj [hunk 1]
```diff
@@ -70,6 +71,7 @@
                              dispatch-token) varr}))]
     (fn dispatch*
       ([]
+        ;; @maz: I'm not sure how to translate strings like this with variable references
        (keys-description (format "Here's what I can %s:" verb) fn-map))
```
Comment (senior) [inline src/metabase/metabot.clj]: You want `(tru "Here's what I can {0}:" verb)`

File: src/metabase/public_settings.clj [hunk 2]
```diff
@@ -9,112 +9,113 @@
...
+  (tru "The base URL of this Metabase instance, e.g. \"http://metabase.my-company.com\".")
   :setter (fn [new-value]
             (setting/set-string! :site-url (when new-value
                                              (cond->> (s/replace new-value #"/$" "")
                                                (not (s/starts-with? new-value "http")) (str "http://"))))))
 
 (defsetting site-locale
-  "The default language for this Metabase instance. This only applies to emails, Pulses, etc. Users' browsers will
-   specify the language used in the user interface."
+  (tru "The default language for this Metabase instance. This only applies to emails, Pulses, etc. Users' browsers will
```
Comment (senior) [inline src/metabase/public_settings.clj]: Gettext can't follow line breaks 😿. In this case, it would be easy to rewrite it like this:
```
(str (tru "The default language for this Metabase instance.")
     (tru "This only applies to emails, Pulses, etc. Users' browsers will specify the language used in the user interface.")
```
That also has the added benefit of creating two internationalized sentences. If one of the sentences shows up somewhere else, we'll only need to translate it once.

**Review by senior — CHANGES_REQUESTED** [review]
Some of the indentation is off, but it's on the multi-line internationalized strings which will need to be fixed up anyway. Not a huge deal, but it's preferable have sentences be in their own internationalization clause so that they can be reused without be re-translated.

File: src/metabase/metabot.clj (hunk 1 above)
Comment (mazameli) [reply src/metabase/metabot.clj]: Okay, so how about something like the one on line 79:
```
(format "I don't know how to %s `%s`.\n%s"`
```
Does that become this? 
```
(tru "I don't know how to {0} `{1}`.\n{2}"`
```

**Comment by mazameli** [comment]
Okay, I made another pass through the two files you left comments on, @senior.

Two things I'm unsure about:
1. If this is the right way to require both `tru` and `trs` (I'm guessing it's not, because this looks redundant):
```
[puppetlabs.i18n.core :as i18n :refer [tru]]
[puppetlabs.i18n.core :as i18n :refer [trs]]
```
2. If I'm using `trs` correctly to translate log messages.

Mind taking another quick look? Thanks!

File: src/metabase/public_settings.clj [hunk 3]
```diff
@@ -9,112 +9,106 @@
             [metabase.public-settings.metastore :as metastore]
             [metabase.util.i18n :refer [available-locales-with-names set-locale]]
             [metabase.util.password :as password]
+            [puppetlabs.i18n.core :as i18n :refer [tru]]
```
Comment (senior) [inline src/metabase/public_settings.clj]: You can remove the :as i18n part, so you'd just have
`[puppetlabs.i18n.core :refer [tru]]`

File: src/metabase/metabot.clj [hunk 4]
```diff
@@ -22,11 +22,13 @@
              [permissions-group :as perms-group]
              [setting :as setting :refer [defsetting]]]
             [metabase.util.urls :as urls]
+            [puppetlabs.i18n.core :as i18n :refer [tru]]
+            [puppetlabs.i18n.core :as i18n :refer [trs]]
```
Comment (senior) [inline src/metabase/metabot.clj]: You can combine these like:
` [puppetlabs.i18n.core :refer [tru trs]]`

File: src/metabase/email.clj [hunk 5]
```diff
@@ -5,20 +5,27 @@
             [postal
              [core :as postal]
              [support :refer [make-props]]]
+            [puppetlabs.i18n.core :as i18n :refer [tru]]
```
Comment (senior) [inline src/metabase/email.clj]: You can remove the `:as i18n` here too

File: src/metabase/metabot.clj [hunk 6]
```diff
@@ -104,26 +106,26 @@
   [& _]
   (let [cards (with-metabot-permissions
                 (filterv mi/can-read? (db/select [Card :id :name :dataset_query], {:order-by [[:id :desc]], :limit 20})))]
-    (str "Here's your " (count cards) " most recent cards:\n" (format-cards cards))))
+    (tru "Here's your " (count cards) " most recent cards:\n" (format-cards cards))))
```
Comment (senior) [inline src/metabase/metabot.clj]: You'll need to switch this to the parameterized '{0}` syntax, i.e. 
``` clojure
(tru "Here's your {0} most recent cards:\n{1}" (count cards) (format-cards cards))
```

File: src/metabase/metabot.clj [hunk 7]
```diff
@@ -104,26 +106,26 @@
...
 
 (defn- card-with-name [card-name]
   (first (u/prog1 (db/select [Card :id :name], :%lower.name [:like (str \% (str/lower-case card-name) \%)])
            (when (> (count <>) 1)
-             (throw (Exception. (str "Could you be a little more specific? I found these cards with names that matched:\n"
+             (throw (Exception. (tru "Could you be a little more specific? I found these cards with names that matched:\n"
```
Comment (senior) [inline src/metabase/metabot.clj]: This also needs the parameterized `{0}` part

File: src/metabase/metabot.clj [hunk 8]
```diff
@@ -160,7 +162,7 @@
 
 
 (def ^:private kanye-quotes
-  (delay (log/debug "Loading kanye quotes...")
+  (delay (log/debug (trs "Loading Kanye quotes..."))
```
Comment (senior) [inline src/metabase/metabot.clj]: 👍 This is using the system locale, which is what you want for logs.

**Review by senior — CHANGES_REQUESTED** [review]
Looking good, just a few minor things to fix

File: src/metabase/metabot.clj (hunk 6 above)
Comment (mazameli) [reply src/metabase/metabot.clj]: Oooohhh, gotcha. I stupidly thought those `\n`s were newlines. 😆

File: src/metabase/public_settings.clj [hunk 9]
```diff
@@ -9,112 +9,106 @@
...
+  (tru "Metabase will cache all saved questions with an average query execution time longer than this many seconds:")
   :type    :integer
   :default 60)
 
 (defsetting query-caching-ttl-ratio
-  "To determine how long each saved question's cached result should stick around, we take the
-   query's average execution time and multiply that by whatever you input here. So if a query
-   takes on average 2 minutes to run, and you input 10 for your multiplier, its cache entry
-   will persist for 20 minutes."
+  (tru "To determine how long each saved question's cached result should stick around, we take the query's average execution time and multiply that by whatever you input here. So if a query takes on average 2 minutes to run, and you input 10 for your multiplier, its cache entry will persist for 20 minutes.")
```
Comment (camsaul) [inline src/metabase/public_settings.clj]: @senior what's the best way we can translate strings like this but maintain a sane line length?

Do we need to do
```clojure
(tru (str "..."
          "..."))
```

?

@mazameli taking something like this that was previously broken out into multiple lines and making it one single massive line makes the code hard to read here on GH. It would be nice if we could keep things broken out somehow so you don't need to scroll horizontally for weeks to read our code

Comment (senior) [reply src/metabase/public_settings.clj]: I agree on making it hard to view in Github, but AFAIK, that's just the way gettext works. It can't follow multiple string literals. This is ok:
```
(str (tru "My first sentence.")
     (tru "My second sentence."))
;;-> Works, creates 2 separate extracted strings
```
But it will silently only extract the first string in the below:
```
(tru (str "My first sentence."
          "My second sentence."))
;; -> Will miss "My second sentence"
```
I tired to generate the translation files from the project to get an example to paste in but that doesn't appear to be working. @tlrobinson how were you generating those when you were testing?

Comment (camsaul) [reply src/metabase/public_settings.clj]: Oh yeah, I forgot about that. Breaking something long up into multiple strings to translate sounds silly. I guess super-long lines will have to do

Comment (tlrobinson) [reply src/metabase/public_settings.clj]: @senior IIRC the general workflow is:

1. Tag strings
2. Run `./bin/i18n/update-translation`
3. Translate `.po` files
4. Run `./bin/i18n/build-translation-resources`

**Comment by mazameli** [comment]
@senior this branch still isn't wanting to pass tests even after I made the fixes we talked about in Slack. The test error in CI is saying `Could not locate metabase/sample_dataset/generate__init.class or metabase/sample_dataset/generate.clj on classpath. Please check that namespaces with dashes use underscores in the Clojure file name.`. Any ideas?

**Comment by senior** [comment]
@mazameli I'm not sure what's going on there. We're trying to run our lein bikeshed plugin, but it can't find our sample dataset code.

@camsaul Do you know what's going on here? The sample dataset code is not in our src directory and it looks like bikeshed can't find it. No idea why we're only seeing it in this branch, seems like we should be seeing it everywhere.

**Comment by camsaul** [comment]
@senior Let me take a look.

File: src/metabase/metabot.clj [hunk 10]
```diff
@@ -104,26 +105,26 @@
...
   (let [cards (with-metabot-permissions
                 (filterv mi/can-read? (db/select [Card :id :name :dataset_query], {:order-by [[:id :desc]], :limit 20})))]
-    (str "Here's your " (count cards) " most recent cards:\n" (format-cards cards))))
+    (tru "Here's your {0} most recent cards:\n{1}" (count cards) (format-cards cards))))
 
 (defn- card-with-name [card-name]
   (first (u/prog1 (db/select [Card :id :name], :%lower.name [:like (str \% (str/lower-case card-name) \%)])
            (when (> (count <>) 1)
-             (throw (Exception. (str "Could you be a little more specific? I found these cards with names that matched:\n"
+             (throw (Exception. (tru "Could you be a little more specific? I found these cards with names that matched:\n{0}"
```
Comment (camsaul) [inline src/metabase/metabot.clj]: `^String`

File: src/metabase/metabot.clj [hunk 11]
```diff
@@ -104,26 +105,26 @@
...
                                      (format-cards <>))))))))
 
 (defn- id-or-name->card [card-id-or-name]
   (cond
     (integer? card-id-or-name)     (db/select-one [Card :id :name], :id card-id-or-name)
     (or (string? card-id-or-name)
         (symbol? card-id-or-name)) (card-with-name card-id-or-name)
-    :else                          (throw (Exception. (format "I don't know what Card `%s` is. Give me a Card ID or name." card-id-or-name)))))
+    :else                          (throw (Exception. (tru "I don't know what Card `{0}` is. Give me a Card ID or name." card-id-or-name)))))
```
Comment (camsaul) [inline src/metabase/metabot.clj]: `^String`

File: src/metabase/metabot.clj [hunk 12]
```diff
@@ -133,8 +134,8 @@
                    (slack/post-chat-message! *channel-id*
                                              nil
                                              attachments)))
-       "Ok, just a second...")
-     (throw (Exception. "Not Found"))))
+       (tru "Ok, just a second..."))
+     (throw (Exception. (tru "Not Found")))))
```
Comment (camsaul) [inline src/metabase/metabot.clj]: `^String`

File: src/metabase/integrations/ldap.clj [hunk 13]
```diff
@@ -8,68 +8,69 @@
              [setting :as setting :refer [defsetting]]
              [user :as user :refer [User]]]
             [metabase.util :as u]
+            [puppetlabs.i18n.core :as i18n :refer [tru]]
```
Comment (camsaul) [inline src/metabase/integrations/ldap.clj]: No need to do `:as i18n` here because you aren't using `i18n` for anything. If we did something like `(i18n/tru ...)` this would make sense but we're not

**Comment by camsaul** [comment]
I've narrowed this down to the changes in `ldap.clj`. Reverting that file makes `bikeshed` pass

**Comment by camsaul** [comment]
@mazameli @senior oh I figured this out. The cannot find file stuff if just a warning and you can ignore that. If you read the whole output this it reveals why it's failing:

> Checking for lines longer than 205 characters.
Badly formatted files:
/home/ubuntu/metabase/src/metabase/public_settings.clj:101:  (tru "To determine how long each saved question's cached result should stick around, we take the query's average execution time and multiply that by whatever you input here. So if a query takes on average 2 minutes to run, and you input 10 for your multiplier, its cache entry will persist for 20 minutes.")

**Comment by camsaul** [comment]
I went ahead and fixed the merged conflicts and squashed this branch so make sure you force-pull when continuing to work on this. The long line needs to be fixed and heads up you're also going to see some reflection linter failures too.

File: src/metabase/public_settings.clj (hunk 9 above)
Comment (camsaul) [reply src/metabase/public_settings.clj]: ^ BTW This is the long line in question

File: src/metabase/metabot.clj (hunk 10 above)
Comment (mazameli) [reply src/metabase/metabot.clj]: Sorry, @camsaul, I don't know what you mean by this comment (or the other two places you've commented this in this file). Do you want me to change something here?

Comment (camsaul) [reply src/metabase/metabot.clj]: Oh yeah I meant that for @senior. 

@senior did we decide type hinting with `^String` doesn't work?

Comment (camsaul) [reply src/metabase/metabot.clj]: @mazameli I went ahead and fixed these.

So basically the `tru` macro doesn't tell people it's returning a string. And certain things like `(Exception.)` need to know they're getting a string to work properly. So we have to tell Clojure that the output of `tru` is a string since it's not doing it itself.

The easiest way I think is just to wrap that in a call to `(str)` which *does* advertise the fact that it returns a string.

So TL;DR do

```clojure
(Exception. (str (tru "...")))`
```

instead of

```clojure
(Exception. (tru "..."))
```

In almost all other places you don't need to wrap things in `(str)` so you can just continue to do those the normal way. If there's another place you need to do it the linter will catch it for you

**Comment by mazameli** [comment]
Yay, it passed tests! 🎉  @senior and @kdoh, could you guys give this one last look over for me pretty please? 🙏  Let's Get This Merged™.

**Review by senior — APPROVED** [review]

**Review by kdoh — APPROVED** [review]
